---
title: 'Repertoire Profiler Report'
date: ''
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
params:
  nb_input: -1
  nb_seq_aligned: -1
  nb_seq_not_aligned: -1
  nb_productive: -1
  nb_unproductive: -1
  nb_clone_assigned: -1
  nb_failed_clone: -1
  constant_rep: ""
  vj_rep: ""
  itol_subscription: "TRUE"
---

\n\n<br><br>\n\n




[comment]: <> (The following script is used to be able to zoom on images)

<script>
  function toggleFullscreen(img) {
    if (img.classList.contains('fullscreen')) {
      img.classList.remove('fullscreen');
    } else {
      document.querySelectorAll('.fullscreen').forEach(el => el.classList.remove('fullscreen'));
      img.classList.add('fullscreen');
    }
  }
</script>
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Images cliquables</title>
    <style>
        .image-container {
            display: flex;
            justify-content: space-around;
            gap: 10px;
        }
        .image-container img {
            width: 33%;
            cursor: pointer;
            transition: transform 0.3s ease;
        }
        .fullscreen {
            position: fixed;
            top: 0;
            left: 0;
            width: 100vw !important;
            height: 100vh !important;
            object-fit: contain;
            background: rgba(0, 0, 0, 0.9);
            z-index: 1000;
            transform: scale(1);
        }
    </style>
</head>

[comment]: <> (End of script to zoom on images)

[comment]: <> (The following script is meant to display the tsv files in a readable way)

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(readr)
library(knitr)

read_tsv_with_dummy <- function(file_path) {
  if (file.info(file_path)$size == 0) {
    return("File is empty")
  } else {
    df <- read_tsv(file_path, show_col_types = FALSE)
    truncated <- FALSE
    # If the file is empty but has column names, add an empty line to make the column names readable
    if (nrow(df) == 0) {
      dummy_row <- as.data.frame(matrix("", nrow = 1, ncol = ncol(df)))
      colnames(dummy_row) <- colnames(df)
      df <- rbind(df, dummy_row)
    }else if (nrow(df) > 10) {
      df <- df[1:10, ]
      truncated <- TRUE
    }
    attr(df, "truncated") <- truncated
    return(df)
  }
}
```

<style>
details.inline {display:inline-block;}
</style>

[comment]: <> (End tsv display)



### Initial dataset

Number of initial input sequences: ```r params$nb_input```

<br><br>

### Igblast sequence annotation


Succes number of sequences: ```r params$nb_seq_aligned```

Failed number of sequences: ```r params$nb_seq_not_aligned```

Total: ```r params$nb_seq_aligned + params$nb_seq_not_aligned```

Success ratio: ```r base::round(params$nb_seq_aligned / (params$nb_seq_aligned + params$nb_seq_not_aligned), 2)```

See the igblast_seq.tsv and the failed_igblast_seq.tsv files in the [tsv](./tsv) folder, as well as their [column description](https://github.com/gael-millot/repertoire_profiler?tab=readme-ov-file#output). 

<br>

### Productive sequences

**Productive sequences** are sequences with the following properties in their VDJ region:<br><ul><li>an open reading frame (start and stop codon).<br></li><li>no defect in the start codon, splicing sites or regulatory elements.<br></li><li>no internal stop codons.<br></li><li>in-frame junction between regions.<br></ul>

Number of productive sequences: ```r params$nb_productive``` &nbsp; <details class="inline"><summary style="color: #1a73e8; font-family: Arial, sans-serif; font-size: 1em; font-weight: bold;">view</summary>
<span style="color:green">*productive_seq.tsv file in the [tsv](./tsv) folder*</span>
```{r, echo=FALSE, results='asis'}
df <- read_tsv_with_dummy("tsv/productive_seq.tsv")
if (is.data.frame(df) && attr(df, "truncated")) {
  cat("\n**(Note: Table truncated to 10 rows. Full file available in the productive_seq.tsv file.)**\n\n")
}
kable(df, format = "markdown", row.names = FALSE)
```
</details>
<br>
Number of unproductive sequences: ```r params$nb_unproductive``` &nbsp; <details class="inline"><summary style="color: #1a73e8; font-family: Arial, sans-serif; font-size: 1em; font-weight: bold;">view</summary>
<span style="color:green">*failed_productive_seq.tsv file in the [tsv](./tsv) folder*</span>
```{r, echo=FALSE, results='asis'}
df <- read_tsv_with_dummy("tsv/failed_productive_seq.tsv")
if (is.data.frame(df) && attr(df, "truncated")) {
  cat("\n**(Note: Table truncated to 10 rows. Full file available in the failed_productive_seq.tsv file.)**\n\n")
}
kable(df, format = "markdown", row.names = FALSE)
```
</details>
<br>

Productiveness ratio: ```r base::round(params$nb_productive / (params$nb_productive + params$nb_unproductive), 2)```

See the productive_seq.tsv and the failed_productive_seq.tsv files in the [tsv](./tsv) folder, as well as their [column description](https://github.com/gael-millot/repertoire_profiler?tab=readme-ov-file#output).

<br>


### Repertoire summary

<br>

Compilation of the [productive_seq.tsv](./tsv) file (and also of the [clone_assigned_seq.tsv](./tsv) file for the donut plots).

<br>

#### Donut plots

<br>
Click on a figure to enlarge. Reclick to go back to the report.
<br>
Only genes are displayed here. See the [donuts.pdf](./pdf/donuts.pdf) file in the [pdf](./pdf) folder for the complete results, including the alleles (different copies of a same gene) proportions.
<br>
Warning: only the first annotation is considered if several are presents in the v_call, j_call or c_call column of the productive_seq.tsv file.

<br>

<div class="image-container">
  <img src="vj_gene_all_donutchart.png" onclick="toggleFullscreen(this)" />
  <img src="vj_gene_annotated_donutchart.png" onclick="toggleFullscreen(this)" />
  <img src="vj_gene_tree_donutchart.png" onclick="toggleFullscreen(this)" />
</div>
<div class="image-container">
  <img src="c_gene_all_donutchart.png" onclick="toggleFullscreen(this)" />
  <img src="c_gene_annotated_donutchart.png" onclick="toggleFullscreen(this)" />
  <img src="c_gene_tree_donutchart.png" onclick="toggleFullscreen(this)" />
</div>

<br><br><br>

#### Heatmaps

<br>
Click on a figure to enlarge. Reclick to go back to the report.
<br>
Only genes are displayed here. See the [gene_repertoire.pdf](./pdf/gene_repertoire.pdf) and [allele_repertoire.pdf](./pdf/allele_repertoire.pdf) files in the [pdf](./pdf) folder for the complete results, including the alleles (different copies of a same gene) counts.
<br>
Warning: only the first annotation is considered if several are presents in the v_call, j_call or c_call column of the productive_seq.tsv file.

<br>

<div class="image-container"> 
  <img src="`r params$constant_rep`" onclick="toggleFullscreen(this)" style="border: 2px solid black; width: 25%;"/> 
  <img src="`r params$vj_rep`" onclick="toggleFullscreen(this)" style="border: 2px solid black; width: 25%"/> 
</div>

<br>



<br><br><br>

###  Clonal prediction 

<br>

**Clone-assigned sequences** are productive sequences with additionnal features:<br><ul><li>germline clustering (clone ID).<br></li><li>germline sequence inference.<br></li><li>mutation load.<br></ul> 

Number of clone-assigned sequences: ```r params$nb_clone_assigned```&nbsp; <details class="inline"><summary style="color: #1a73e8; font-family: Arial, sans-serif; font-size: 1em; font-weight: bold;">view</summary>
<span style="color:green">*clone_assigned_seq.tsv file in the [tsv](./tsv) folder*</span>
```{r, echo=FALSE, results='asis'}
df <- read_tsv_with_dummy("tsv/clone_assigned_seq.tsv")
if (is.data.frame(df) && attr(df, "truncated")) {
  cat("\n**(Note: Table truncated to 10 rows. Full file available in the clone_assigned_seq.tsv file.)**\n\n")
}
kable(df, format = "markdown", row.names = FALSE)
```
</details>
```{r include = FALSE}
nb_failed_clone <- ifelse(params$nb_failed_clone > 0, params$nb_failed_clone, 0)
```
<br>

Number of failed clone-assigned or germline sequences: ```r nb_failed_clone```

<br>
Passing ratio: ```r base::round(params$nb_clone_assigned / (params$nb_clone_assigned + nb_failed_clone), 2)```

See the clone_assigned_seq.tsv, as well as both the failed_clone_assigned_seq.tsv and the failed_clonal_germline.tsv (sequences from the productive file that fail to be clone-assigned by `DefineClones.py` and `CreateGermlines.py`) files in the [tsv](./tsv) folder, as well as their [column description](https://github.com/gael-millot/repertoire_profiler?tab=readme-ov-file#output).

<br>
Number of sequences per clone ID: &nbsp; <details class="inline"><summary style="color: #1a73e8; font-family: Arial, sans-serif; font-size: 1em; font-weight: bold;">view</summary>
<span style="color:green">*clone_id_count.tsv file in the [tsv](./tsv) folder*</span>
```{r, echo=FALSE, results='asis'}
df <- read_tsv_with_dummy("tsv/clone_id_count.tsv")
if (is.data.frame(df) && attr(df, "truncated")) {
  cat("\n**(Note: Table truncated to 10 rows. Full file available in the clone_id_count.tsv file.)**\n\n")
}
kable(df, format = "markdown", row.names = FALSE)
```
</details>
<br>


### Sequence alignments

<br>

**Sequence Alignments** are available in the [alignments/nuc](./alignments/nuc) folder for nucleotide alignments and in the [alignments/aa](./alignments/aa) folder for amino acids alignments. Such alignments are performed using [Abalign](https://academic.oup.com/nar/article/51/W1/W17/7173809) (alignments guided by IMGT antibody numbering schemes).<br>Several options are possible to visualize and save as picture:<br>
<ul>
<li>Click on the html file to see the alignment.<br>
<ul>
<li>Warning: use the horizontal scrollbar of the window + the right click of the mouse on the alignment to see the left part of the representation.</li><br>
<li>To change the aligments colors ("Nucleotide" recommanded for nucleotides):<br><br><br>
<div class="image-container">
<img src="doc_images/color.png" onclick="toggleFullscreen(this)" />
</div>
</li><br>
<li>To annotate the FR and CDR regions, use the .gff file in the same folder than the opened .html file:<br><br><br>
<div class="image-container">
<img src="doc_images/gff.png" onclick="toggleFullscreen(this)" />
</div>
</li><br>
<li>Schematic illustration of these regions:<br><br><br>
<div class="image-container">
<img src="doc_images/FR_CDR.png" onclick="toggleFullscreen(this)" />
</div>
<div class="image-container">
<img src="doc_images/collier_de_perle.png" onclick="toggleFullscreen(this)" />
</div>
</li><br>
<li>Export as picture does not work well as part of the sequence not displayed are not exported (screenshot).</li><br><br><br>
</ul>
</li>
<li>In [Jalview](https://www.jalview.org/download/).<br>
<ul>
<li>import the .fasta file with the same name as the .html file:<br><br><br>
<div class="image-container">
<img src="doc_images/jalview_open.png" onclick="toggleFullscreen(this)" />
</div>
</li><br>
<li>Add colors:<br><br><br>
<div class="image-container">
<img src="doc_images/jalview_color.png" onclick="toggleFullscreen(this)" />
</div>
</li><br><li>Set a threshold to see the differences:<br><br><br>
<div class="image-container">
<img src="doc_images/jalview_diff.png" onclick="toggleFullscreen(this)" />
</div>
</li><br><li>A high threshold set all the columns with at least one difference in white:<br><br><br>
<div class="image-container">
<img src="doc_images/jalview_diff_threshold_high.png" onclick="toggleFullscreen(this)" />
</div>
</li><br><li>A low threshold set all the different minority events in each column in white:<br><br><br>
<div class="image-container">
<img src="doc_images/jalview_diff_threshold_low.png" onclick="toggleFullscreen(this)" />
</div>
</li><br><li>To export as .png for instance:<br><br><br>
<div class="image-container">
<img src="doc_images/jalview_export.png" onclick="toggleFullscreen(this)" />
</div>
</li>
</ul>
</li>
</ul>
<br><br><br>

###  Phylogeny
<br>

`r  if(params$itol_subscription == "TRUE"){"Open the different **sequences_\\<SEQ_NAME\\>_align.fasta_itol_url.txt** files in the [phylo/aa](./phylo/aa) folder and click on the link to see the trees."}`
`r  if(params$itol_subscription == "FALSE"){"To see the trees, follow the link: https://itol.embl.de/upload.cgi. Choose a name for your tree (optional) and upload a '*.treefile' file from the [phylo/aa](./phylo/aa) folder. The trees will then be displayed by iTOL."}`

Alignment visualization is available for each clonal group used in each tree, in the [phylo/aa](./phylo/aa) folder in html format.
Tree scale is the number of mutations per position. Multiply by the length of the sequence to have the number of mutations per sequence


<br><br>

###  Backup


See the [reports](./reports) folder for all the details of the analysis, including:<br><ul><li>**nextflow.config** Parameter settings.<br></li><li>**Run_info.txt** Variables.<br></li><li>**trace.txt** and **timeline.html** Date of execution, CPU% and Memory requirement for each steps.<br></ul>

```{r include = FALSE}
#remove work directory from current directory
wd <- getwd()
parts <- unlist(strsplit(wd, "/"))
path <- paste(parts[1:(length(parts) - 3)], collapse = "/")
path <- paste0(path, "/")
```
Warning: the full .nextflow.log is in: ```r path```. Indeed, the one in the [reports](./reports) folder is not complete (miss the end).


<br><br>

<style>
  #TOC { max-width: 200px !important;}
  /* .tocify-item { font-size: 100% !important; } */
</style>

<style> 
  body .container, body .container-fluid { max-width: 90% !important; width: 100% !important; margin: 0 auto;  } 
</style>

<style>
  .toc-content.col-xs-12.col-sm-8.col-md-9 { margin-left: 10px !important; padding-left: 0 !important; } 
</style>

<style> /* Force a fixated width, float left and add space on the right */ .col-xs-12.col-sm-4.col-md-3 { width: 250px !important; float: left; margin-right: 20px; /* reduce interior margin */ padding: 5px; } </style>
