---
title: 'Repertoire Profiler Report'
date: ''
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
params:
  nb_input: -1
  nb_seq_aligned: -1
  nb_seq_not_aligned: -1
  nb_productive: -1
  nb_unproductive: -1
  nb_clone_assigned: -1
  nb_failed_clone: -1
  constant_rep: ""
  vj_rep: ""
---

\n\n<br /><br />\n\n




[comment]: <> (The following script is used to be able to zoom on images)

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(knitr)
```

<script>
  function toggleFullscreen(img) {
    if (img.classList.contains('fullscreen')) {
      img.classList.remove('fullscreen');
    } else {
      document.querySelectorAll('.fullscreen').forEach(el => el.classList.remove('fullscreen'));
      img.classList.add('fullscreen');
    }
  }
</script>
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Images cliquables</title>
    <style>
        .image-container {
            display: flex;
            justify-content: space-around;
            gap: 10px;
        }
        .image-container img {
            width: 33%;
            cursor: pointer;
            transition: transform 0.3s ease;
        }
        .fullscreen {
            position: fixed;
            top: 0;
            left: 0;
            width: 100vw !important;
            height: 100vh !important;
            object-fit: contain;
            background: rgba(0, 0, 0, 0.9);
            z-index: 1000;
            transform: scale(1);
        }
    </style>
</head>

[comment]: <> (End of script to zoom on images)



### Initial dataset

Number of initial input sequences : ```r params$nb_input```

<br><br>

### Igblast sequence annotation

[Succes](files/igblast_aligned_seq.tsv) number of sequences : ```r params$nb_seq_aligned```

[Failed](files/igblast_unaligned_seq.tsv) number of sequences : ```r params$nb_seq_not_aligned```

Total : ```r params$nb_seq_aligned + params$nb_seq_not_aligned```

Success ratio : ```r base::round(params$nb_seq_aligned / (params$nb_seq_aligned + params$nb_seq_not_aligned), 2)```

<br>

### Productive sequences

**Productive sequences** are sequences with: (1) coding region with an open reading frame, (2) no defect in the start codon, splicing sites or regulatory elements, (3) no internal stop codons, (4) an in-frame junction region. <br>

Number of [productive sequences](files/productive_seq.tsv) : ```r params$nb_productive```

Number of [unproductive sequences](files/unproductive_seq.tsv) : ```r params$nb_unproductive```

Productiveness ratio : ```r base::round(params$nb_productive / (params$nb_productive + params$nb_unproductive), 2)```

<br>


### Repertoire summary

<br>

#### Donut plots

<br>


<div class="image-container">
  <img src="png/vj_allele_all_donutchart.png" onclick="toggleFullscreen(this)" />
  <img src="png/vj_allele_annotated_donutchart.png" onclick="toggleFullscreen(this)" />
  <img src="png/vj_allele_tree_donutchart.png" onclick="toggleFullscreen(this)" />
</div>

<br><br>

See also the [donuts.pdf](files/donuts.pdf) file in the files folder.

<br><br><br>

#### Heatmaps

<br>

<div style="display: flex; justify-content: space-around;"> 
  <img src="png/`r params$constant_rep`" width="40%" style="border: 2px solid black;" /> 
  <img src="png/`r params$vj_rep`" width="40%" style="border: 2px solid black;" /> 
</div>

<br />

See the [gene_repertoire.pdf](files/gene_repertoire.pdf) and [allele_repertoire.pdf](files/allele_repertoire.pdf) files in the files folder.

<br><br><br>

###  Clonal prediction 

<br>

**Clone-assigned sequences** are productive sequences with additionnal features: germline clustering (clone ID), allele reannotation (germinal_v_call and germinal_j_call columns), mutation load, distance and sequence nickname (annotation from the metadata file) added. <br>
Some sequences from the productive file can **fail to be clone-assigned**, and will not be in the clone_assigned_seq.tsv file. 

Number of [clone-assigned sequences](files/clone_assigned_seq.tsv) : ```r params$nb_clone_assigned```

Number of [failed clone-assigned sequences](files/non_clone_assigned_sequence.tsv) : ```r params$nb_failed_clone```

Passing ratio : ```r base::round(params$nb_clone_assigned / (params$nb_clone_assigned + params$nb_failed_clone), 2)```

<br>

See the [germ_tree.pdf](files/germ_tree.pdf) file in the files folder.
<br>

See also:<br /><ul><li>[germ_tree_seq.tsv](files/germ_tree_seq.tsv)<br /></li><li>[germ_tree_dismissed_seq.tsv](files/germ_tree_dismissed_seq.tsv)<br /></li><li>[germ_tree_clone_id.tsv](files/germ_tree_clone_id.tsv)<br /></li><li>[germ_tree_dismissed_clone_id.tsv](files/germ_tree_dismissed_clone_id.tsv)


<br><br><br>

###  Amino acid sequences phylogeny 

<br>

**Only available for heavy chain analysis.**

Open the different sequences_\<SEQ_NAME\>_align.fasta_itol_url.txt files in the [phylo](./phylo) folder and click on the link to see the trees.

<br><br><br>

###  Backup


See the [reports](./reports) folder for all the details of the analysis, including the parameters used in the .config file


Full .nextflow.log is in: /mnt/c/Users/Gael/Documents/Git_projects/14985_loot<br />The one in the [reports](./reports) folder is not complete (miss the end)


<br /><br />

<style>
  #TOC { max-width: 200px !important;}
  /* .tocify-item { font-size: 100% !important; } */
</style>

<style> 
  body .container, body .container-fluid { max-width: 90% !important; width: 100% !important; margin: 0 auto;  } 
</style>

<style>
  .toc-content.col-xs-12.col-sm-8.col-md-9 { margin-left: 10px !important; padding-left: 0 !important; } 
</style>

<style> /* Forcer une largeur fixe, flotter à gauche et ajouter un espace à droite */ .col-xs-12.col-sm-4.col-md-3 { width: 250px !important; float: left; margin-right: 20px; /* éventuellement, réduire sa marge intérieure si besoin */ padding: 5px; } </style>
