---
title: 'Repertoire Profiler Report'
date: ''
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
params:
    igblast_organism: ""
    igblast_loci: ""
    igblast_B_heavy_chain: "TRUE"
    igblast_B_lambda_chain: "TRUE"
    igblast_B_kappa_chain: "TRUE"
    igblast_T_alpha_chain: "TRUE"
    igblast_T_beta_chain: "TRUE"
    igblast_T_gamma_chain: "TRUE"
    igblast_T_delta_chain: "TRUE"
    nb_input: -1
    nb_igblast: -1
    nb_unigblast: -1
    nb_productive: -1
    nb_unproductive: -1
    nb_wanted: -1
    nb_unwanted: -1
    nb_dist_ignored: -1
    nb_clone_tot: -1
    nb_unclone_tot: -1
    nb_clone_unassignment: -1
    nb_clone_ungermline: -1
    clone_distance: -1
    constant_rep: ""
    vj_rep: ""
    align_soft: ""
    itol_subscription: "TRUE"
    warning_collect: ""
---

\n\n<br><br>\n\n




[comment 1]: <> (The following script is used to be able to zoom on images)

<script>
  function toggleFullscreen(img) {
    if (img.classList.contains('fullscreen')) {
      img.classList.remove('fullscreen');
    } else {
      document.querySelectorAll('.fullscreen').forEach(el => el.classList.remove('fullscreen'));
      img.classList.add('fullscreen');
    }
  }
</script>
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Images cliquables</title>
    <style>
        .image-container {
            display: flex;
            justify-content: space-around;
            gap: 10px;
        }
        .image-container img {
            width: 33%;
            cursor: pointer;
            transition: transform 0.3s ease;
        }
        .fullscreen {
            position: fixed;
            top: 0;
            left: 0;
            width: 100vw !important;
            height: 100vh !important;
            object-fit: contain;
            background: rgba(0, 0, 0, 0.9);
            z-index: 1000;
            transform: scale(1);
        }
    </style>
</head>

[comment 2]: <> (End of script to zoom on images)

[comment 3]: <> (The following script is meant to display the tsv files in a readable way)

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

read_tsv_with_dummy <- function(file_path){
    if( ! file.exists(file_path)){
        stop(paste0("\n\n================\n\nINTERNAL ERROR IN Print_report PROCESS.\nfile_path DOES NOT EXISTS: ", file_path, "\n\nPLEASE, REPORT AN ISSUE HERE https://gitlab.pasteur.fr/gmillot/repertoire_profiler/-/issues OR AT gael.millot<AT>pasteur.fr.\n\n========\n\n"), call. = FALSE)
    }else if(file.info(file_path)$size == 0){
        stop(paste0("\n\n================\n\nINTERNAL ERROR IN Print_report PROCESS.\nfile_path FILE IS EMPTY BUT SHOULD NOT BE READ HERE. PATH IS: ", file_path, "\n\nPLEASE, REPORT AN ISSUE HERE https://gitlab.pasteur.fr/gmillot/repertoire_profiler/-/issues OR AT gael.millot<AT>pasteur.fr.\n\n========\n\n"), call. = FALSE)
    }else{
        df <- readr::read_tsv(file_path, show_col_types = FALSE)
        truncated <- FALSE
        # If the file is empty but has column names, add an empty line to make the column names readable
        if (nrow(df) == 0){
            dummy_row <- as.data.frame(matrix("", nrow = 1, ncol = ncol(df)))
            colnames(dummy_row) <- colnames(df)
            df <- rbind(df, dummy_row)
        }else if(nrow(df) > 10){
            df <- df[1:10, ]
            truncated <- TRUE
        }
        return(list(df = df, truncated = truncated))
    }
}
```

<style>
details.inline {display:inline-block;}
</style>

[comment 4]: <> (End tsv display)

[comment 5]: <> (Openable .config file)

```{r echo=FALSE, results='hide'}
txt <- readLines("./config_file.txt", warn = FALSE)
# basic HTML-escaping so config symbols display correctly
escape <- function(x) {
  x <- gsub("&", "&amp;", x, fixed = TRUE)
  x <- gsub("<", "&lt;",  x, fixed = TRUE)
  x <- gsub(">", "&gt;",  x, fixed = TRUE)
  x
}
out <- c(
  "<!doctype html>",
  "<meta charset='utf-8'>",
  "<title>nextflow.config</title>",
  "<pre style='white-space:pre; font-family:monospace;'>",
  escape(txt),
  "</pre>"
)
writeLines(out, "nextflow.config.html")
``` 

[comment 6]: <> (end Openable .config file)

[comment 7]: <> (assignation of params, for later reassignation)

```{r, echo=FALSE, results='asis'}
    igblast_organism <- params$igblast_organism
    igblast_loci <- params$igblast_loci
    igblast_B_heavy_chain <- params$igblast_B_heavy_chain
    igblast_B_lambda_chain <- params$igblast_B_lambda_chain
    igblast_B_kappa_chain <- params$igblast_B_kappa_chain
    igblast_T_alpha_chain <- params$igblast_T_alpha_chain
    igblast_T_beta_chain <- params$igblast_T_beta_chain
    igblast_T_gamma_chain <- params$igblast_T_gamma_chain
    igblast_T_delta_chain <- params$igblast_T_delta_chain
    nb_input <- params$nb_input
    nb_igblast <- params$nb_igblast
    nb_unigblast <- params$nb_unigblast
    nb_productive <- params$nb_productive
    nb_unproductive <- params$nb_unproductive
    nb_wanted <- params$nb_wanted
    nb_unwanted <- params$nb_unwanted
    nb_dist_ignored <- params$nb_dist_ignored
    nb_clone_tot <- params$nb_clone_tot
    nb_unclone_tot <- params$nb_unclone_tot
    nb_clone_unassignment <- params$nb_clone_unassignment
    nb_clone_ungermline <- params$nb_clone_ungermline
    clone_distance <- params$clone_distance
    constant_rep <- params$constant_rep
    vj_rep <- params$vj_rep
    align_soft <- params$align_soft
    itol_subscription <- params$itol_subscription
    warning_collect <- params$warning_collect
```

[comment 8]: <> (end assignation of params, for later reassignation)


```{r, echo=FALSE, results='asis'}
if(nb_productive == -1){
    if(nb_wanted != -1){
        stop(paste0("\n\n================\n\nINTERNAL ERROR IN Print_report PROCESS.\nnb_wanted CANNOT BE OTHER THAN -1 IF nb_productive IS -1:\nnb_wanted: ", nb_wanted, "\nnb_productive: ", nb_productive, "\n\nPLEASE, REPORT AN ISSUE HERE https://gitlab.pasteur.fr/gmillot/repertoire_profiler/-/issues OR AT gael.millot<AT>pasteur.fr.\n\n========\n\n"), call. = FALSE)
    }
    if(nb_clone_tot != -1){
        stop(paste0("\n\n================\n\nINTERNAL ERROR IN Print_report PROCESS.\nnb_clone_tot CANNOT BE OTHER THAN -1 IF nb_productive IS -1:\nnb_clone_tot: ", nb_clone_tot, "\nnb_productive: ", nb_productive, "\n\nPLEASE, REPORT AN ISSUE HERE https://gitlab.pasteur.fr/gmillot/repertoire_profiler/-/issues OR AT gael.millot<AT>pasteur.fr.\n\n========\n\n"), call. = FALSE)
    }
}
if(nb_wanted == -1){
    if(nb_clone_tot != -1){
        stop(paste0("\n\n================\n\nINTERNAL ERROR IN Print_report PROCESS.\nnb_clone_tot CANNOT BE OTHER THAN -1 IF nb_wanted IS -1:\nnb_clone_tot: ", nb_clone_tot, "\nnb_wanted: ", b_wanted, "\n\nPLEASE, REPORT AN ISSUE HERE https://gitlab.pasteur.fr/gmillot/repertoire_profiler/-/issues OR AT gael.millot<AT>pasteur.fr.\n\n========\n\n"), call. = FALSE)
    }
}
```
Description of the folders and files can be found [here](https://github.com/gael-millot/repertoire_profiler?tab=readme-ov-file#output)
<br>
The <i>[nextflow.config](./reports/nextflow.config.html)</i> file mentioned here is the one gathering all the parameters set for the present analysis. A copy is saved in the [reports](./reports) folder.
<br><br>

### Initial dataset

Number of initial input sequences: ```r nb_input```

<br><br>

### Sequence annotation using Igblast


Number of annotated sequences: ```r nb_igblast```
<br>
Number of sequences failing annotation: ```r nb_unigblast```
<br>
Total: ```r nb_igblast + nb_unigblast```

Success ratio: ```r base::round(nb_igblast / (nb_igblast + nb_unigblast), 2)```

See the igblast_seq.tsv and the failed_igblast_seq.tsv files in the [tsv](./tsv) folder, as well as their [column description](https://github.com/gael-millot/repertoire_profiler?tab=readme-ov-file#output). 

<br><br>

### Wanted sequences


**Productive sequences** are igblast annotated sequences with the following properties in their VDJ region:<br><ul><li>an open reading frame (start and stop codon).<br></li><li>no defect in the start codon, splicing sites or regulatory elements.<br></li><li>no internal stop codons.<br></li><li>in-frame junction between regions.<br></ul>
Unproductive sequences could deserve a visual inspection of the sequence profile to exclude any problem of automatic profile translation.
<br><br>

**Wanted sequences** are productive sequences selected according to the parameters of the <i>nextflow.config</i> file:<br><ul><li><code>igblast_organism</code>:&nbsp;&nbsp;&nbsp;`r igblast_organism`<br></li><li><code>igblast_loci</code>:&nbsp;&nbsp;&nbsp;`r igblast_loci`<br></li><li><code>igblast_B_heavy_chain</code>:&nbsp;&nbsp;&nbsp;`r igblast_B_heavy_chain`<br></li><li><code>igblast_B_lambda_chain</code>:&nbsp;&nbsp;&nbsp;`r igblast_B_lambda_chain`<br></li><li><code>igblast_B_kappa_chain</code>:&nbsp;&nbsp;&nbsp;`r igblast_B_kappa_chain`<br></li><li><code>igblast_T_alpha_chain</code>:&nbsp;&nbsp;&nbsp;`r igblast_T_alpha_chain`<br></li><li><code>igblast_T_beta_chain</code>:&nbsp;&nbsp;&nbsp;`r igblast_T_beta_chain`<br></li><li><code>igblast_T_gamma_chain</code>:&nbsp;&nbsp;&nbsp;`r igblast_T_gamma_chain`<br></li><li><code>igblast_T_delta_chain</code>:&nbsp;&nbsp;&nbsp;`r igblast_T_delta_chain`<br></ul>
<br>

```{r, echo=FALSE, results='asis'}
if(nb_productive != -1){
    if(nb_wanted == -1){
        nb_wanted <- 0
    }
    if(nb_unwanted == -1){
        nb_unwanted <- 0
    }
    # warning: cat() concatenates without paste0() but with default argument sep = " "
    cat(paste0('
Number of productive sequences: ', nb_productive
    ))
    if(nb_productive > 0){
        cat(paste0('
&nbsp; <details class="inline"><summary style="color: #1a73e8; font-family: Arial, sans-serif; font-size: 1em; font-weight: bold;">view</summary>
<span style="color:green">*productive_seq.tsv file in the [tsv](./tsv) folder*</span>
        '))
        tempo_read <- read_tsv_with_dummy("tsv/productive_seq.tsv")
        df <- tempo_read$df
        truncated_log <- tempo_read$truncated
        if(is.data.frame(df) && truncated_log){
            cat(paste0('
\n**(Note: Table truncated to 10 rows. Full file available in the unproductive_seq.tsv file.)**\n\n
            '))
            cat(knitr::kable(df, format = "markdown", row.names = FALSE))
        }else{
            cat(knitr::kable(df, format = "html", row.names = FALSE))
        }
        cat(paste0('
</details>
        '))
    }
    cat(paste0('
<br>
Number of unproductive sequences: ', nb_unproductive
    ))
    if(nb_unproductive > 0){
        cat(paste0('
&nbsp; <details class="inline"><summary style="color: #1a73e8; font-family: Arial, sans-serif; font-size: 1em; font-weight: bold;">view</summary>
<span style="color:green">*unproductive_seq.tsv file in the [tsv](./tsv) folder*</span>
        '))
        tempo_read <- read_tsv_with_dummy("tsv/unproductive_seq.tsv")
        df <- tempo_read$df
        truncated_log <- tempo_read$truncated
        if(is.data.frame(df) && truncated_log){
            cat(paste0('
\n**(Note: Table truncated to 10 rows. Full file available in the unproductive_seq.tsv file.)**\n\n
            '))
            cat(knitr::kable(df, format = "markdown", row.names = FALSE))
        }else{
            cat(knitr::kable(df, format = "html", row.names = FALSE))
        }
        cat(paste0('
</details>
        '))
    }
    cat(paste0('
<br>
Number of wanted sequences: ', nb_wanted    
    ))
    if(nb_wanted > 0){
        cat(paste0('
&nbsp; <details class="inline"><summary style="color: #1a73e8; font-family: Arial, sans-serif; font-size: 1em; font-weight: bold;">view</summary>
<span style="color:green">*wanted_seq.tsv file in the [tsv](./tsv) folder*</span>
        '))
        tempo_read <- read_tsv_with_dummy("tsv/wanted_seq.tsv")
        df <- tempo_read$df
        truncated_log <- tempo_read$truncated
        if(is.data.frame(df) && truncated_log){
            cat(paste0('
\n**(Note: Table truncated to 10 rows. Full file available in the wanted_seq.tsv file.)**\n\n
            '))
            cat(knitr::kable(df, format = "markdown", row.names = FALSE))
        }else{
            cat(knitr::kable(df, format = "html", row.names = FALSE))
        }
        cat(paste0('
</details>
        '))
    }
    cat(paste0('
<br>
Number of unwanted sequences: ', nb_unwanted
    ))
    if(nb_unwanted > 0){
        cat(paste0('
&nbsp; <details class="inline"><summary style="color: #1a73e8; font-family: Arial, sans-serif; font-size: 1em; font-weight: bold;">view</summary>
<span style="color:green">*unwanted_seq.tsv file in the [tsv](./tsv) folder*</span>
        '))
        tempo_read <- read_tsv_with_dummy("tsv/unwanted_seq.tsv")
        df <- tempo_read$df
        truncated_log <- tempo_read$truncated
        if(is.data.frame(df) && truncated_log){
            cat(paste0('
\n**(Note: Table truncated to 10 rows. Full file available in the unwanted_seq.tsv file.)**\n\n
            '))
            cat(knitr::kable(df, format = "markdown", row.names = FALSE))
        }else{
            cat(knitr::kable(df, format = "html", row.names = FALSE))
        }
        cat(paste0('
</details>
        '))
    }
    cat(paste0('
<br>
Total (wanted + unproductive + unwanted): ', nb_wanted + nb_unproductive + nb_unwanted, '

Productive / Total ratio: ', base::round(nb_productive / (nb_wanted + nb_unproductive + nb_unwanted), 2), '

Wanted / Productive ratio: ', base::round(nb_wanted / nb_productive, 2), '

See the productive_seq.tsv, unproductive_seq.tsv, wanted_seq.tsv and unwanted_seq.tsv files in the [tsv](./tsv) folder, as well as their [column description](https://github.com/gael-millot/repertoire_profiler?tab=readme-ov-file#output).

    '))
}else{
    cat(paste0('
No productive sequence detected.
    '))
}
```

<br><br>

### Repertoire summary

```{r, echo=FALSE, results='asis'}
if(nb_wanted > 0){
    cat(paste0('

Compilation of the [wanted_seq.tsv](./tsv) file.
<br>
Click on a figure to enlarge. Reclick to go back to the report.
<br>
Only genes are displayed here. See the pdf files in the [pdf](./pdf) folder for the complete results, including the alleles (different copies of a same gene) counts.
<br>
Warning: only the first annotation is considered if several are presents in the v_call, j_call or c_call column of the wanted_seq.tsv file.

<br>

#### Heatmaps

<br>

'))
    cat('<div class="image-container">')
    if(file.exists(constant_rep)){
        # warning: cat() concatenates without paste0() but with default argument sep = " "
        cat(paste0('<img src="', constant_rep, '" onclick="toggleFullscreen(this)" ',
  'style="border:1px solid black;width:25%;display:inline-block;margin:0;" />'))
    }
    if(file.exists(vj_rep)){
        cat(paste0('<img src="', vj_rep, '" onclick="toggleFullscreen(this)" ',
  'style="border:1px solid black;width:25%;display:inline-block;margin:0;" />'))
    }
    cat('</div><br><br><br>')
    cat(paste0('

#### Donut plots

<br>

'))
    cat('<div class="image-container">')
    if(file.exists("vj_gene_all_donutchart.png")){
        cat('<img src="vj_gene_all_donutchart.png" onclick="toggleFullscreen(this)" style="border: 1px solid black"/>')
    }
    if(file.exists("vj_gene_annotated_donutchart.png")){
        cat('<img src="vj_gene_annotated_donutchart.png" onclick="toggleFullscreen(this)" style="border: 1px solid black"/>')
    }
    cat('</div><div class="image-container">')
    if(file.exists("c_gene_all_donutchart.png")){
        cat('<img src="c_gene_all_donutchart.png" onclick="toggleFullscreen(this)" style="border: 1px solid black"/>')
    }
    if(file.exists("c_gene_annotated_donutchart.png")){
        cat('<img src="c_gene_annotated_donutchart.png" onclick="toggleFullscreen(this)" style="border: 1px solid black"/>')
    }
    cat('</div><br><br><br>')
}else{
    cat(paste0('
No wanted sequence detected.
    '))
}
```
<br><br>

###  Clonal prediction

```{r, echo=FALSE, results='asis'}
if(nb_wanted > 0){
    nb_clone_tot <- ifelse(nb_clone_tot > 0, nb_clone_tot, 0)
    nb_unclone_tot <- ifelse(nb_unclone_tot > 0, nb_unclone_tot, 0)
    nb_clone_unassignment <- ifelse(nb_clone_unassignment > 0, nb_clone_unassignment, 0)
    nb_clone_ungermline <- ifelse(nb_clone_ungermline > 0, nb_clone_ungermline, 0)
    nb_dist_ignored <- ifelse(nb_dist_ignored > 0, nb_dist_ignored, 0)
    cat(paste0('

#### Definition

**Clone-assigned sequences** are wanted sequences with additionnal features:<br><ul><li>germline clustering (clone ID).<br></li><li>inference of clonal germline sequence (different from the sequence germline inference).<br></li><li>mutation load.<br></ul><br>
Clonal prediction aims to gather all the sequences that putatively come from a same germline cell. See <a href="https://changeo.readthedocs.io/en/stable/examples/cloning.html#assigning-clones">here</a>, <a href="https://shazam.readthedocs.io/en/stable/vignettes/DistToNearest-Vignette/">here</a> and <a href="https://genomemedicine.biomedcentral.com/articles/10.1186/s13073-015-0243-2">this article</a> for details. In summary, it follows these steps:<br><ul><li>grouping the sequences according to "same V, J and junction length" (to facilitate the distance computation).<br></li><li>for each group, 2x2 distance computation using by default (<code>clone_model</code> and <code>clone_normalize</code> parameters of the <i>nextflow.config</i> file) the <a href="https://biology.stackexchange.com/questions/23523/hamming-distance-between-two-dna-strings">Hamming distance</a>.<br></li><li>cutoff definition.<br></li><li>using the cutoff (<code>clone_distance</code> parameter of the <i>nextflow.config</i> file) to define clonal groups inside each "same V, J and junction length" group.<br></ul><br>

#### Parameter settings


#### Results


Number of clone-assigned sequences: ', nb_clone_tot
    ))
    if(nb_clone_tot > 0){
        cat(paste0('
&nbsp; <details class="inline"><summary style="color: #1a73e8; font-family: Arial, sans-serif; font-size: 1em; font-weight: bold;">view</summary>
<span style="color:green">*clone_assigned_seq.tsv file in the [tsv](./tsv) folder*</span>
        '))
        tempo_read <- read_tsv_with_dummy("tsv/clone_assigned_seq.tsv")
        df <- tempo_read$df
        truncated_log <- tempo_read$truncated
        if(is.data.frame(df) && truncated_log){
            cat(paste0('
\n**(Note: Table truncated to 10 rows. Full file available in the clone_assigned_seq.tsv file.)**\n\n
            '))
            cat(knitr::kable(df, format = "markdown", row.names = FALSE))
        }else{
            cat(knitr::kable(df, format = "html", row.names = FALSE))
        }
        cat(paste0('
</details>
        '))
    }
    cat(paste0('
<br>
Number of sequences with failed clonal assignement: ', nb_clone_unassignment
    ))
    if(nb_clone_unassignment > 0){
        cat(paste0('
&nbsp; <details class="inline"><summary style="color: #1a73e8; font-family: Arial, sans-serif; font-size: 1em; font-weight: bold;">view</summary>
<span style="color:green">*failed_clone_assigned_seq.tsv file in the [tsv](./tsv) folder*</span>
        '))
        tempo_read <- read_tsv_with_dummy("tsv/failed_clone_assigned_seq.tsv")
        df <- tempo_read$df
        truncated_log <- tempo_read$truncated
        if(is.data.frame(df) && truncated_log){
            cat(paste0('
\n**(Note: Table truncated to 10 rows. Full file available in the failed_clone_assigned_seq.tsv file.)**\n\n
            '))
            cat(knitr::kable(df, format = "markdown", row.names = FALSE))
        }else{
            cat(knitr::kable(df, format = "html", row.names = FALSE))
        }
        cat(paste0('
</details>
        '))
    }
    cat(paste0('
<br>
Number of clone assigned sequences with failed germline sequence computation: ', nb_clone_ungermline
    ))
    if(nb_clone_ungermline > 0){
        cat(paste0('
&nbsp; <details class="inline"><summary style="color: #1a73e8; font-family: Arial, sans-serif; font-size: 1em; font-weight: bold;">view</summary>
<span style="color:green">*failed_clonal_germline_seq.tsv file in the [tsv](./tsv) folder*</span>
        '))
        tempo_read <- read_tsv_with_dummy("tsv/failed_clonal_germline_seq.tsv")
        df <- tempo_read$df
        truncated_log <- tempo_read$truncated
        if(is.data.frame(df) && truncated_log){
            cat(paste0('
\n**(Note: Table truncated to 10 rows. Full file available in the failed_clonal_germline_seq.tsv file.)**\n\n
            '))
            cat(knitr::kable(df, format = "markdown", row.names = FALSE))
        }else{
            cat(knitr::kable(df, format = "html", row.names = FALSE))
        }
        cat(paste0('
</details>
        '))
    }
    cat(paste0('
<br>
Total (clone-assigned + failed_clone_assignment + failed_clonal_germline): ', nb_clone_tot + nb_clone_unassignment + nb_clone_ungermline, '

Clone assigned ratio: ', base::round(nb_clone_tot / (nb_clone_tot + nb_clone_unassignment + nb_clone_ungermline), 2), '

See the clone_assigned_seq.tsv, as well as both the failed_clone_assigned_seq.tsv and the failed_clonal_germline_seq.tsv (sequences from the wanted file that fail to be clone-assigned by `DefineClones.py` and `CreateGermlines.py`) files in the [tsv](./tsv) folder, as well as their [column description](https://github.com/gael-millot/repertoire_profiler?tab=readme-ov-file#output).
    '))
    if(nb_clone_tot > 0){
        cat(paste0('
<br>
Number of sequences per clone ID: &nbsp; <details class="inline"><summary style="color: #1a73e8; font-family: Arial, sans-serif; font-size: 1em; font-weight: bold;">view</summary>
<span style="color:green">*clone_id_count.tsv file in the [tsv](./tsv) folder*</span>
        '))
        tempo_read <- read_tsv_with_dummy("tsv/clone_id_count.tsv")
        df <- tempo_read$df
        truncated_log <- tempo_read$truncated
        if(is.data.frame(df) && truncated_log){
            cat(paste0('
\n**(Note: Table truncated to 10 rows. Full file available in the clone_id_count.tsv file.)**\n\n
            '))
            cat(knitr::kable(df, format = "markdown", row.names = FALSE))
        }else{
            cat(knitr::kable(df, format = "html", row.names = FALSE))
        }
        cat(paste0('
</details>
<br>

#### Donut plots

Compilation of the [clone_assigned_seq.tsv](./tsv) file.
<br>
Click on a figure to enlarge. Reclick to go back to the report.
<br>
Only genes are displayed here. See the [donuts.pdf](./pdf/donuts.pdf) file in the [pdf](./pdf) folder for the complete results, including the alleles (different copies of a same gene) proportions.
<br>
Warning: only the first annotation is considered if several are presents in the v_call, j_call or c_call column of the wanted_seq.tsv file.

<br>
'))
        cat('<div class="image-container">')
        if(file.exists("vj_gene_clone_donutchart.png")){
            cat('<img src="vj_gene_clone_donutchart.png" onclick="toggleFullscreen(this)" style="border: 1px solid black"/>')
        }
        if(file.exists("c_gene_clone_donutchart.png")){
            cat('<img src="c_gene_clone_donutchart.png" onclick="toggleFullscreen(this)" style="border: 1px solid black"/>')
        }
        cat('</div><br><br><br>')
        cat(paste0('
<br>

#### Threshold distance

Threshold value set in the `clone_distance` parameter of the *nextflow.config* file: ', clone_distance, '

This value is used to assemble sequences into clonal groups.

On the graphs below, the vertical dotted red line (threshold value) must separate the two bumps of the bimodal distributions.

If it is not the case, please adapt the value of the `clone_distance` parameter and rerun.

If no bimodal distribution is observed, it means that the number of sequences analyzed is too low for correct clonal group attribution. A lack of values could be due to a lack of sequences with the same V, J and junction length.

Of note: 

Number of sequences used in the distance matrix (from wanted_seq.tsv): ', nb_wanted, '
<br>
Number of sequences that could not be used in the distance matrix: ', nb_dist_ignored, '

Sequence ratio: ', base::round((nb_wanted - nb_dist_ignored)/ nb_wanted, 2), '

<br>

<div class="image-container">
<img src="seq_distance_manual.png" onclick="toggleFullscreen(this)" style="border: 1px solid black"/>
<img src="seq_distance_mixture_auto.png" onclick="toggleFullscreen(this)" style="border: 1px solid black"/>
<img src="seq_distance_smoothed_auto.png" onclick="toggleFullscreen(this)" style="border: 1px solid black"/>
</div>
        '))
    }
}else{
    cat(paste0('
No clonal assignment obtained.
    '))
}
```

<br><br>

### Sequence alignments

```{r, echo=FALSE, results='asis'}
if(nb_wanted > 0){
    cat(paste0('
Sequence Alignments are available in the [alignments/nuc](./alignments/nuc) folder for nucleotide alignments and in the [alignments/aa](./alignments/aa) folder for amino acids alignments. Click on the desired html files to see the alignments.<br><br>See [here](reports/alignments_viz.html) the ways to deal with alignment visualization and saving.
    '))
    if(align_soft == "abalign"){
        tempo_read <- read_tsv_with_dummy("tsv/failed_abalign_align.tsv")
        df <- tempo_read$df
        truncated_log <- tempo_read$truncated
        cat(paste0('
Such alignments were obtained using [Abalign](https://academic.oup.com/nar/article/51/W1/W17/7173809) (alignments guided by IMGT antibody numbering schemes).<br><br>Warning: some sequences may not be aligned, because they are too short or out of the VDJ region.
<br><br>
Number of failed alignments: <code class="r-output">', nrow(df), '</code>&nbsp; <details class="inline"><summary style="color: #1a73e8; font-family: Arial, sans-serif; font-size: 1em; font-weight: bold;">view</summary>
<span style="color:green">*failed_abalign_align.tsv file in the [tsv](.) folder*</span>
        '))
        if(is.data.frame(df) && truncated_log){
            cat(paste0('
\n**(Note: Table truncated to 10 rows. Full file available in the [tsv](.) folder.)**\n\n
            '))
            cat(knitr::kable(df, format = "markdown", row.names = FALSE))
        }else{
            cat(knitr::kable(df, format = "html", row.names = FALSE))
        }
        cat(paste0('
</details>
<br><br>
        '))
    }else if(align_soft == "mafft"){
        cat(paste0('
Such alignments were obtain using [Mafft](https://mafft.cbrc.jp/alignment/server/index.html).
        '))
    }
    cat(paste0('
Warning in clonal group alignments ([alignments/nuc/clonal](./alignments/nuc/clonal) and [alignments/aa/clonal](./alignments/aa/clonal) folders):<br><ul><li>short sequences sometimes hardly align with the full length germline clonal sequence.<br></li><li>FWR/CDR feature highlighting (.gff files) is not correct at the FWR3 / CDR3 junction, due to the stretch of hyphens and N/X in that region, but coordinates reported in the <i>clonal_germline_fwr3_end</i> and <i>clonal_germline_cdr3_start</i> columns of the <i>clone_assigned_seq.tsv</i> should correctly indicate the TGT (cys) boundary.
    '))
}else{
    cat(paste0('
No wanted sequence detected.
    '))
}
```
<br><br>

###  Phylogeny

```{r, echo=FALSE, results='asis'}
if(nb_wanted > 0){
    #remove work directory from current directory
    if(length(list.files("phylo/aa", all.files = TRUE, no.. = TRUE)) == 0 && length(list.files("phylo/aa", all.files = TRUE, no.. = TRUE)) == 0){
        cat(paste0('
No tree computed because of lack of sequences in clonal groups.
        '))
    }else{
        if(itol_subscription == "TRUE"){
            cat(paste0('
Open the different **sequences_\\<SEQ_NAME\\>_align.fasta_itol_url.txt** files in the [phylo/aa](./phylo/aa) folder and click on the link to see the trees.\n\n
            '))
        }else{
            cat(paste0('
To see the trees, follow the link: https://itol.embl.de/upload.cgi. Choose a name for your tree (optional) and upload a "*.treefile" file from the [phylo/aa](./phylo/aa) folder. The trees will then be displayed by iTOL.\n\n
            '))
        }
        cat(paste0('
Alignment visualization is available for each clonal group used in each tree, in the [phylo/aa](./phylo/aa) folder in html format.\nTree scale is the number of mutations per position. Multiply by the length of the sequence to have the number of mutations per sequence.
        '))
    }
}else{
    cat(paste0('
No wanted sequence detected.
    '))
}
```


<br><br>

###  Warnings recapitulation

```{r, echo=FALSE, results='asis'}
#remove work directory from current directory
if(length(warning_collect) == 1 && all(warning_collect == "")){
    cat(paste0('No warning message reported.'))
}else{
     # remove strings
    
    warning_collect <- warning_collect[warning_collect != ""]
    warning_collect <- gsub(x = warning_collect, pattern = "\n", replacement = "<br>")
    warning_collect <- gsub(x = warning_collect, pattern = "^WARNING:.*", replacement = "\n")
    warning_collect <- paste0(warning_collect, collapse = "<br>") # single string
    warning_collect <- gsub(x = warning_collect, pattern = "\n", replacement = "")
    warning_collect <-  unique(warning_collect) #vector of string
    # warning_collect <- warning_collect[ ! grepl(x = warning_collect, pattern = "^ *WARNING: *")]
    # end remove strings
    cat(warning_collect)
}
```


<br><br>

###  Backup

See the [reports](./reports) folder for all the details of the analysis, including:<br><ul><li>**nextflow.config** Parameter settings.<br></li><li>**Run_info.txt** Variables.<br></li><li>**trace.txt** and **timeline.html** Date of execution, CPU% and Memory requirement for each step.<br></ul>

```{r include = FALSE}
#remove work directory from current directory
wd <- getwd()
parts <- unlist(strsplit(wd, "/"))
path <- paste(parts[1:(length(parts) - 3)], collapse = "/")
path <- paste0(path, "/")
```
Warning: the full .nextflow.log is in: ```r path```. Indeed, the one in the [reports](./reports) folder is not complete (miss the end).


<br><br>

<style>
  #TOC { max-width: 200px !important;}
  /* .tocify-item { font-size: 100% !important; } */
</style>

<style> 
  body .container, body .container-fluid { max-width: 90% !important; width: 100% !important; margin: 0 auto;  } 
</style>

<style>
  .toc-content.col-xs-12.col-sm-8.col-md-9 { margin-left: 10px !important; padding-left: 0 !important; } 
</style>

<style> /* Force a fixated width, float left and add space on the right */ .col-xs-12.col-sm-4.col-md-3 { width: 250px !important; float: left; margin-right: 20px; /* reduce interior margin */ padding: 5px; } </style>
